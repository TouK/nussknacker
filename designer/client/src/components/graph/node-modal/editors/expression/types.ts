export enum ExpressionLang {
    SQL = "sql",
    SpEL = "spel",
    SpELTemplate = "spelTemplate",
    String = "string",
    JSON = "json",
    TabularDataDefinition = "tabularDataDefinition",
}

export type ExpressionObj = {
    expression: string;
    language: ExpressionLang | string;
};

export enum EditorMode {
    SpEL = "spel",
    SpELTemplate = "spelTemplate",
    SQL = "sql",
}
