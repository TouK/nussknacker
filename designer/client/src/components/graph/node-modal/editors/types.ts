import { TypingResult } from "../../../../types";
import { EditorType } from "./expression/Editor";
import { EditorMode, ExpressionLang } from "./expression/types";

//TODO: FIXME
export type ParamType = {
    name?: string;
    typ?: TypingResult;
    editor?: {
        type: `${EditorType}`;
    };
    editors?: {
        type: `${EditorType}`;
        language: ExpressionLang | string;
        dictId?: string;
    }[];
    defaultValue: {
        language: EditorMode | string;
        expression: string;
    };
    additionalVariables?: Record<string, unknown>;
    variablesToHide?: unknown[];
    branchParam?: boolean;
    hintText?: string | null;
    label?: string;
    requiredParam?: boolean;
};
