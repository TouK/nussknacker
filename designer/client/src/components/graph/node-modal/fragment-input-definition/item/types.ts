import { Expression, ReturnedType } from "../../../../../types";
import { resolveRefClazzName } from "./utils";

export type onChangeType = string | number | boolean | FixedValuesOption | FixedValuesOption[] | ValueCompileTimeValidation | ValueEditor;

export interface ValueCompileTimeValidation {
    validationExpression: Expression;
    validationFailedMessage?: string;
}

export interface FragmentValidation {
    valueCompileTimeValidation: ValueCompileTimeValidation | null;
}

export enum FixedValuesType {
    "ValueInputWithFixedValuesProvided" = "ValueInputWithFixedValuesProvided",
    "ValueInputWithDictEditor" = "ValueInputWithDictEditor",
}

export enum InputMode {
    "FixedList" = "InputModeFixedList",
    "AnyValueWithSuggestions" = "InputModeAnyWithSuggestions",
    "AnyValue" = "InputModeAny",
}

export interface FixedValuesOption {
    expression: string;
    label: string;
    icon?: string;
}

export interface GenericParameterVariant {
    uuid: string;
    required: boolean;
    name: string;
    typ?: ReturnedType;
    initialValue: FixedValuesOption | null;
    hintText: string | null;
    //It's only to satisfy typescript
    expression?: Expression;
}

export interface ValueEditor {
    type: FixedValuesType;
    fixedValuesList: FixedValuesOption[] | null;
    allowOtherValue: boolean | null;
    dictId: string;
}

export interface DefaultParameterVariant extends GenericParameterVariant, FragmentValidation {
    name: string;
    valueEditor: null;
}

export interface FixedListParameterVariant extends GenericParameterVariant, FragmentValidation {
    valueEditor: ValueEditor;
    presetSelection?: string;
}
export interface AnyValueWithSuggestionsParameterVariant extends GenericParameterVariant, FragmentValidation {
    valueEditor: ValueEditor;
    presetSelection?: string;
}
export interface AnyValueParameterVariant extends GenericParameterVariant, FragmentValidation {
    fixedValuesType: FixedValuesType;
    valueEditor: null;
}

export type PermittedTypeParameterVariant = FixedListParameterVariant | AnyValueWithSuggestionsParameterVariant | AnyValueParameterVariant;

export type FragmentInputParameter = PermittedTypeParameterVariant | DefaultParameterVariant;

export function isFixedListParameter(item: PermittedTypeParameterVariant): item is FixedListParameterVariant {
    return item.valueEditor?.allowOtherValue === false;
}

export function isAnyValueWithSuggestionsParameter(item: PermittedTypeParameterVariant): item is AnyValueWithSuggestionsParameterVariant {
    return item?.valueEditor?.allowOtherValue === true;
}

export function isAnyValueParameter(item: PermittedTypeParameterVariant): item is AnyValueParameterVariant {
    return item.valueEditor === null;
}

export function isPermittedTypeVariant(item: FragmentInputParameter): item is PermittedTypeParameterVariant {
    return [
        resolveRefClazzName(item.typ.refClazzName) === "String",
        resolveRefClazzName(item.typ.refClazzName) === "Boolean",
        resolveRefClazzName(item.typ.refClazzName) === "Long",
        resolveRefClazzName(item.typ.refClazzName) === "Integer",
    ].includes(true);
}

export type FieldName = `$param.${string}.$${string}`;
