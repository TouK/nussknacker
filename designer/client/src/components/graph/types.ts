import { ProcessCounts } from "../../reducers/graph";
import {
    injectNode,
    Layout,
    layoutChanged,
    nodeAdded,
    nodesConnected,
    nodesDisconnected,
    resetSelection,
    toggleSelection,
} from "../../actions/nk";
import { Capabilities } from "../../reducers/selectors/other";
import { Scenario } from "../Process/types";

type ScenarioGraphProps = {
    nodesConnected: typeof nodesConnected;
    nodesDisconnected: typeof nodesDisconnected;
    layoutChanged: typeof layoutChanged;
    injectNode: typeof injectNode;
    nodeAdded: typeof nodeAdded;
    resetSelection: typeof resetSelection;
    toggleSelection: typeof toggleSelection;

    scenario: Scenario;
    divId: string;
    nodeIdPrefixForFragmentTests?: string;
    processCounts: ProcessCounts;
    capabilities: Capabilities;
    layout: Layout;

    readonly?: boolean;
    nodeSelectionEnabled?: boolean;
    isDraggingOver?: boolean;
    isFragment?: false | null;

    connectDropTarget;
};

type FragmentGraphProps = {
    scenario: Scenario;
    divId: string;
    nodeIdPrefixForFragmentTests: string;
    processCounts: ProcessCounts;
    layout: Layout;
    isFragment: true;
    readonly: true;
};

export type GraphProps = ScenarioGraphProps | FragmentGraphProps;

export enum Events {
    LINK_CONNECT = "link:connect",
    LINK_DISCONNECT = "link:disconnect",
    LINK_MOUSEOVER = "link:mouseover",
    LINK_MOUSEENTER = "link:mouseenter",
    LINK_MOUSELEAVE = "link:mouseleave",
    LINK_POINTERDOWN = "link:pointerdown",
    CELL_POINTERUP = "cell:pointerup",
    CELL_POINTERDOWN = "cell:pointerdown",
    CELL_POINTERMOVE = "cell:pointermove",
    CELL_POINTERCLICK = "cell:pointerclick",
    CELL_POINTERDBLCLICK = "cell:pointerdblclick",
    CELL_MOUSEOVER = "cell:mouseover",
    CELL_MOUSEOUT = "cell:mouseout",
    CELL_MOVED = "cellCustom:moved",
    BLANK_POINTERDOWN = "blank:pointerdown",
    BLANK_POINTERUP = "blank:pointerup",
    BLANK_POINTERMOVE = "blank:pointermove",
    BLANK_MOUSEOVER = "blank:mouseover",
    REMOVE = "remove",
    ADD = "add",
    CHANGE_POSITION = "change:position",
    CHANGE_DRAG_OVER = "change:draggedOver",
}
