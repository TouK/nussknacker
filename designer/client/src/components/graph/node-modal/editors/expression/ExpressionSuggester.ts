import HttpService from "../../../../../http/HttpService";
import { ExpressionLang } from "./types";
import { VariableTypes } from "../../../../../types";

export type CaretPosition2d = { row: number; column: number };

export type ExpressionSuggestion = {
    methodName: string;
    refClazz: { display: string };
    fromClass: boolean;
    description?: string;
    parameters?: { name: string; refClazz: { display: string } }[];
};

export interface ExpressionSuggester {
    suggestionsFor(inputValue: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]>;
}

export class BackendExpressionSuggester implements ExpressionSuggester {
    constructor(
        private language: ExpressionLang | string,
        private variableTypes: VariableTypes,
        private processingType: string,
        private httpService: typeof HttpService,
    ) {}

    suggestionsFor = (inputValue: string, caretPosition2d: CaretPosition2d): Promise<ExpressionSuggestion[]> =>
        this.httpService
            .getExpressionSuggestions(this.processingType, {
                expression: {
                    language: this.language,
                    expression: inputValue,
                },
                caretPosition2d,
                variableTypes: this.variableTypes,
            })
            .then((response) => response.data);
}
