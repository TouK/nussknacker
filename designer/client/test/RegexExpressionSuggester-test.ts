import {RegexExpressionSuggester} from "../src/components/graph/node-modal/editors/expression/ExpressionSuggester";
import {map} from "lodash";
import {expect, describe, it} from '@jest/globals';
import {ClassDefinition} from "../src/types";

const typesInformation: ClassDefinition[] = [
  {
    "clazzName": {"refClazzName": "org.A", type:"", display: "", fields: {}, params: []},
    "methods": {"fooString": {"refClazz": {"refClazzName": "java.lang.String"} }, "barB": {"refClazz": {"refClazzName": "org.B"} } }
  },
  {
    "clazzName": {"refClazzName": "org.B", type:"", display: "", fields: {}, params: []},
    "methods": {"bazC": {"refClazz": {"refClazzName": "org.C"} } }
  },
  {
    "clazzName": {"refClazzName": "org.C", type:"", display: "", fields: {}, params: []},
    "methods": {"quaxString": {"refClazz": {"refClazzName": "java.lang.String"} }}
  },
  {
    "clazzName": {"refClazzName": "org.AA", type:"", display: "", fields: {}, params: []},
    "methods": {"fooString": {"refClazz": {"refClazzName": "java.lang.String"} }, "barB": {"refClazz": {"refClazzName": "org.C"} } }
  },
  {
    "clazzName": {"refClazzName": "org.WithList", type:"", display: "", fields: {}, params: []},
    "methods": {"listField": {"refClazz": {"refClazzName": "java.util.List", params: [{refClazzName: "org.A"}]} }}
  },
  {
    "clazzName": {"refClazzName": "java.lang.String", type:"", display: "", fields: {}, params: []},
    "methods": {"toUpperCase": {"refClazz": {"refClazzName": "java.lang.String"} }}
  },
  {
    "clazzName": {"refClazzName": "java.util.LocalDateTime", type:"", display: "", fields: {}, params: []},
    "methods": {"isBefore": {"refClazz": {"refClazzName": "java.lang.Boolean"}, "params": {"name": "arg0", "refClazz": "java.util.LocalDateTime"}}}
  },
  {
    "clazzName": {"refClazzName": "org.Util", type:"", display: "", fields: {}, params: []},
    "methods": {"now": {"refClazz": {"refClazzName": "java.util.LocalDateTime"}}}
  }
];

const variables = {
  "input": {refClazzName: "org.A"},
  "other": {refClazzName: "org.C"},
  "ANOTHER": {refClazzName: "org.A"},
  "dynamicMap": {refClazzName: "java.util.Map", fields: {'intField': {refClazzName: 'java.lang.Integer'}, 'aField': {refClazzName: "org.A"}} },
  "listVar": {refClazzName: "org.WithList" },
  "util": {refClazzName: "org.Util"},
  "union": {union: [
      {refClazzName: "org.A"},
      {refClazzName: "org.B"},
      {refClazzName: "org.AA"}]},
  "unionOfLists": {union: [
      {"refClazz": "java.util.List", params: [{refClazzName: "org.A"}]},
      {"refClazz": "java.util.List", params: [{refClazzName: "org.B"}]}
    ]},
  "dict": {dict: {
    id: "fooDict",
    valueType: {refClazzName: "org.A"}
  }}
};

const suggestionsFor = (inputValue, caretPosition2d?, stubbedDictSuggestions?) => {
  const _caretPosition2d = caretPosition2d ? caretPosition2d : { row: 0, column: inputValue.length }
  const stubService = {
    fetchDictLabelSuggestions(processingType, dictId, labelPattern): Promise<any> {
      return new Promise(resolve => resolve({ data: stubbedDictSuggestions }))
    }
  }
  const expressionSuggester = new RegexExpressionSuggester(typesInformation, variables, "fooProcessingType", stubService)
  return expressionSuggester.suggestionsFor(inputValue, _caretPosition2d)
}

describe("expression suggester", () => {
  it("should not suggest anything for empty input", () => {
    return expect(suggestionsFor("")).resolves.toMatchSnapshot()
  })

  it("should suggest all global variables if # specified", () => {
    return expect(suggestionsFor("#")).resolves.toMatchSnapshot()
  })

  it("should suggest all global variables if # specified (multiline)", () => {
    return expect(suggestionsFor(`asdasd\n
    #\n
    dsadasdas`, {row: 1, column: 1})).resolves.toMatchSnapshot()
  })

  it("should filter global variables suggestions", () => {
    return expect(suggestionsFor("#ot")).resolves.toMatchSnapshot()
  })

  it("should filter uppercase global variables suggestions", () => {
    return expect(suggestionsFor("#ANO")).resolves.toMatchSnapshot()
  })

  it("should suggest global variable", () => {
    return expect(suggestionsFor("#inpu")).resolves.toMatchSnapshot()
  })

  it("should suggest global variable methods", () => {
    return expect(suggestionsFor("#input.")).resolves.toMatchSnapshot()
  })

  it("should suggest dict variable methods", () => {
    let stubbedDictSuggestions = [
      { key: "one", label: "One" },
      { key: "two", label: "Two" }
    ]
    return expect(suggestionsFor("#dict.", null, stubbedDictSuggestions)).resolves.toMatchSnapshot()
  })


  it("should suggest dict variable methods using indexer syntax", () => {
    let stubbedDictSuggestions = [
      { key: "sentence-with-spaces-and-dots", label: "Sentence with spaces and . dots" }
    ]
    const correctInputs = ["#dict['", "#dict['S", "#dict['Sentence w", "#dict['Sentence with spaces and . dots"]
    map(correctInputs, inputValue => {
      return expect(suggestionsFor(inputValue, null, stubbedDictSuggestions)).resolves.toMatchSnapshot()
    })
  })

  it("should suggest filtered global variable methods", () => {
    return expect(suggestionsFor("#input.fo")).resolves.toMatchSnapshot()
  })

  it("should suggest filtered global variable methods based not on beginning of the method", () => {
    return expect(suggestionsFor("#input.string")).resolves.toMatchSnapshot()
  })

  it("should suggest methods for object returned from method", () => {
    return expect(suggestionsFor("#input.barB.bazC.")).resolves.toMatchSnapshot()
  })

  it("should suggest methods for union objects", () => {
    return expect(suggestionsFor("#union.")).resolves.toMatchSnapshot()
  })

  it("should suggest methods for object returned from method from union objects", () => {
    return expect(suggestionsFor("#union.bazC.")).resolves.toMatchSnapshot()
  })

  it("should suggest in complex expression #1", () => {
    return expect(suggestionsFor("#input.foo + #input.barB.bazC.quax", {row: 0, column: "#input.foo".length })).resolves.toMatchSnapshot()
  })

  it("should suggest in complex expression #2", () => {
    return expect(suggestionsFor("#input.foo + #input.barB.bazC.quax")).resolves.toMatchSnapshot()
  })

  it("should suggest in complex expression #3", () => {
    return expect(suggestionsFor("#input.barB.bazC.quaxString.toUp")).resolves.toMatchSnapshot()
  })


  it("should not suggest anything if suggestion already applied with space at the end", () => {
    return expect(suggestionsFor("#input.fooString ")).resolves.toMatchSnapshot()
  })

  it("should suggest for invocations with method parameters #1", () => {
    return expect(suggestionsFor("#input.foo + #input.barB.bazC('1').quax")).resolves.toMatchSnapshot()
  })

  it("should suggest for invocations with method parameters #2", () => {
    return expect(suggestionsFor("#input.foo + #input.barB.bazC('1', #input.foo, 2).quax")).resolves.toMatchSnapshot()
  })

  it("should suggest for multiline code #1", () => {
    return expect(suggestionsFor("#input\n.fo", {row: 1, column: ".fo".length })).resolves.toMatchSnapshot()
  })

  it("should suggest for multiline code #2", () => {
    return expect(suggestionsFor("#input\n.barB\n.", {row: 2, column: ".".length })).resolves.toMatchSnapshot()
  })

  it("should suggest for multiline code #3", () => {
    return expect(suggestionsFor("#input\n.ba\n.barC", {row: 1, column: ".ba".length })).resolves.toMatchSnapshot()
  })

  it("should omit whitespace formatting in suggest for multiline code #1", () => {
    return expect(suggestionsFor("#input\n  .ba", {row: 1, column: "  .ba".length })).resolves.toMatchSnapshot()
  })

  it("should omit whitespace formatting in suggest for multiline code #2", () => {
    return expect(suggestionsFor("#input\n  .barB\n  .ba", {row: 2, column: "  .ba".length })).resolves.toMatchSnapshot()
  })

  it("should omit whitespace formatting in suggest for multiline code #3", () => {
    return expect(suggestionsFor("#input\n  .ba\n  .bazC", {row: 1, column: "  .ba".length })).resolves.toMatchSnapshot()
  })

  it("should omit whitespace formatting in suggest for multiline code #4", () => {
    return expect(suggestionsFor("#input\n  .barB.ba", {row: 1, column: "  .barB.ba".length })).resolves.toMatchSnapshot()
  })

  it("should omit whitespace formatting in suggest for multiline code #5", () => {
    return expect(suggestionsFor("#input\n  .barB.bazC\n  .quaxString.", {row: 2, column: "  .quaxString.".length })).resolves.toMatchSnapshot()
  })

  it("should suggest field in typed map", () => {
    return expect(suggestionsFor("#dynamicMap.int")).resolves.toMatchSnapshot()
  })

  it("should suggest embedded field in typed map", () => {
    return expect(suggestionsFor("#dynamicMap.aField.f")).resolves.toMatchSnapshot()
  })

  it("should suggest #this fields in simple projection", () => {
    return expect(suggestionsFor("#listVar.listField.![#this.f]", {row: 0, column: "#listVar.listField.![#this.f".length })).resolves.toMatchSnapshot()
  })

  it("should suggest #this fields in projection on union of lists", () => {
    return expect(suggestionsFor("#unionOfLists.![#this.f]", {row: 0, column: "#unionOfLists.![#this.f".length })).resolves.toMatchSnapshot()
  })

  it("should suggest #this fields in projection after selection", () => {
    return expect(suggestionsFor("#listVar.listField.?[#this == 'value'].![#this.f]", {row: 0, column: "#listVar.listField.?[#this == 'value'].![#this.f".length })).resolves.toMatchSnapshot()
  })

  it("handles negated parameters with projections and selections", () => {
    return expect(suggestionsFor("!#listVar.listField.?[#this == 'value'].![#this.f]", {row: 0, column: "!#listVar.listField.?[#this == 'value'].![#this.f".length })).resolves.toMatchSnapshot()
  })

  it("should support nested method invocations", () => {
    return expect(suggestionsFor("#util.now(#other.quaxString.toUpperCase().)", {row: 0, column: "#util.now(#other.quaxString.toUpperCase().".length })).resolves.toMatchSnapshot()
  })

  it("should support safe navigation", () => {
    return expect(suggestionsFor("#input?.barB.bazC?.")).resolves.toMatchSnapshot()
  })
})


describe("remove finished selections from input", () => {

  const expressionSuggester = new RegexExpressionSuggester(typesInformation, variables, null, null)

  it("leaves unfinished selection", () => {
    const original = "#input.one.two.?[#this."
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual(original)
  })

  it("leaves projections", () => {
    const original = "#input.one.two.![#this.value]"
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual(original)
  })

  it("removes one selection", () => {
    const original = "#input.one.two.?[#this.value == 1].three"
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual("#input.one.two.three")
  })

  it("removes many selections", () => {
    const original = "#input.one.two.?[#this.value == 1].three.?[#this.value == 2].four.?[#this.ala =="
    const removed = expressionSuggester._removeFinishedSelectionFromExpressionPart(original)
    expect(removed).toEqual("#input.one.two.three.four.?[#this.ala ==")
  })

})

describe("normalize multiline input", () => {

  const expressionSuggester = new RegexExpressionSuggester(typesInformation, variables, null, null)

  it("normalize multiline input #1", () => {
    const extracted = expressionSuggester._normalizeMultilineInputToSingleLine("#input\n  .barB.bazC\n  .quaxString.", {row: 1, column: "  .barB.bazC".length })
    expect(extracted).toEqual({
      normalizedInput: "#input.barB.bazC.quaxString.",
      normalizedCaretPosition: "#input.barB.bazC".length
    })
  })

  it("normalize multiline input #2", () => {
    const extracted = expressionSuggester._normalizeMultilineInputToSingleLine("#input\n  .barB.bazC\n  .quaxString.", {row: 2, column: "  .quaxString.".length })
    expect(extracted).toEqual({
      normalizedInput: "#input.barB.bazC.quaxString.",
      normalizedCaretPosition: "#input.barB.bazC.quaxString.".length
    })
  })

})
