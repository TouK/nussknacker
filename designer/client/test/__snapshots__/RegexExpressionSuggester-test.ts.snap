// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression suggester handles negated parameters with projections and selections 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should filter global variables suggestions 1`] = `
[
  {
    "methodName": "#other",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should filter uppercase global variables suggestions 1`] = `
[
  {
    "methodName": "#ANOTHER",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should not suggest anything for empty input 1`] = `[]`;

exports[`expression suggester should not suggest anything if suggestion already applied with space at the end 1`] = `[]`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #1 1`] = `
[
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "B",
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #2 1`] = `
[
  {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #3 1`] = `
[
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "B",
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #4 1`] = `
[
  {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should omit whitespace formatting in suggest for multiline code #5 1`] = `
[
  {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in projection after selection 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in projection on union of lists 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest #this fields in simple projection 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest all global variables if # specified (multiline) 1`] = `
[
  {
    "methodName": "#input",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
  {
    "methodName": "#other",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
  {
    "methodName": "#ANOTHER",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
  {
    "methodName": "#dynamicMap",
    "refClazz": {
      "display": "Map",
      "fields": {
        "aField": {
          "display": "A",
          "refClazzName": "org.A",
        },
        "intField": {
          "display": "Integer",
          "refClazzName": "java.lang.Integer",
        },
      },
      "refClazzName": "java.util.Map",
    },
  },
  {
    "methodName": "#listVar",
    "refClazz": {
      "display": "WithList",
      "refClazzName": "org.WithList",
    },
  },
  {
    "methodName": "#util",
    "refClazz": {
      "display": "Util",
      "refClazzName": "org.Util",
    },
  },
  {
    "methodName": "#union",
    "refClazz": {
      "union": [
        {
          "display": "A",
          "refClazzName": "org.A",
        },
        {
          "display": "B",
          "refClazzName": "org.B",
        },
        {
          "display": "AA",
          "refClazzName": "org.AA",
        },
      ],
    },
  },
  {
    "methodName": "#unionOfLists",
    "refClazz": {
      "union": [
        {
          "params": [
            {
              "display": "A",
              "refClazzName": "org.A",
            },
          ],
          "refClazz": "java.util.List",
        },
        {
          "params": [
            {
              "display": "B",
              "refClazzName": "org.B",
            },
          ],
          "refClazz": "java.util.List",
        },
      ],
    },
  },
  {
    "methodName": "#dict",
    "refClazz": {
      "dict": {
        "id": "fooDict",
        "valueType": {
          "display": "A",
          "refClazzName": "org.A",
        },
      },
    },
  },
]
`;

exports[`expression suggester should suggest all global variables if # specified 1`] = `
[
  {
    "methodName": "#input",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
  {
    "methodName": "#other",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
  {
    "methodName": "#ANOTHER",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
  {
    "methodName": "#dynamicMap",
    "refClazz": {
      "display": "Map",
      "fields": {
        "aField": {
          "display": "A",
          "refClazzName": "org.A",
        },
        "intField": {
          "display": "Integer",
          "refClazzName": "java.lang.Integer",
        },
      },
      "refClazzName": "java.util.Map",
    },
  },
  {
    "methodName": "#listVar",
    "refClazz": {
      "display": "WithList",
      "refClazzName": "org.WithList",
    },
  },
  {
    "methodName": "#util",
    "refClazz": {
      "display": "Util",
      "refClazzName": "org.Util",
    },
  },
  {
    "methodName": "#union",
    "refClazz": {
      "union": [
        {
          "display": "A",
          "refClazzName": "org.A",
        },
        {
          "display": "B",
          "refClazzName": "org.B",
        },
        {
          "display": "AA",
          "refClazzName": "org.AA",
        },
      ],
    },
  },
  {
    "methodName": "#unionOfLists",
    "refClazz": {
      "union": [
        {
          "params": [
            {
              "display": "A",
              "refClazzName": "org.A",
            },
          ],
          "refClazz": "java.util.List",
        },
        {
          "params": [
            {
              "display": "B",
              "refClazzName": "org.B",
            },
          ],
          "refClazz": "java.util.List",
        },
      ],
    },
  },
  {
    "methodName": "#dict",
    "refClazz": {
      "dict": {
        "id": "fooDict",
        "valueType": {
          "display": "A",
          "refClazzName": "org.A",
        },
      },
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods 1`] = `
[
  {
    "fromClass": false,
    "methodName": "One",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
  {
    "fromClass": false,
    "methodName": "Two",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 1`] = `
[
  {
    "fromClass": false,
    "methodName": "Sentence with spaces and . dots",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 2`] = `
[
  {
    "fromClass": false,
    "methodName": "Sentence with spaces and . dots",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 3`] = `
[
  {
    "fromClass": false,
    "methodName": "Sentence with spaces and . dots",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest dict variable methods using indexer syntax 4`] = `
[
  {
    "fromClass": false,
    "methodName": "Sentence with spaces and . dots",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest embedded field in typed map 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest field in typed map 1`] = `
[
  {
    "methodName": "intField",
    "refClazz": {
      "display": "Integer",
      "refClazzName": "java.lang.Integer",
    },
  },
]
`;

exports[`expression suggester should suggest filtered global variable methods 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest filtered global variable methods based not on beginning of the method 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for invocations with method parameters #1 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for invocations with method parameters #2 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #1 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #2 1`] = `
[
  {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should suggest for multiline code #3 1`] = `
[
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "B",
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should suggest global variable 1`] = `
[
  {
    "methodName": "#input",
    "refClazz": {
      "display": "A",
      "refClazzName": "org.A",
    },
  },
]
`;

exports[`expression suggester should suggest global variable methods 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "B",
      "refClazzName": "org.B",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #1 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #2 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest in complex expression #3 1`] = `
[
  {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for object returned from method 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for object returned from method from union objects 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should suggest methods for union objects 1`] = `
[
  {
    "fromClass": false,
    "methodName": "fooString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "B",
      "refClazzName": "org.B",
    },
  },
  {
    "fromClass": false,
    "methodName": "bazC",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
  {
    "fromClass": false,
    "methodName": "barB",
    "refClazz": {
      "display": "C",
      "refClazzName": "org.C",
    },
  },
]
`;

exports[`expression suggester should support nested method invocations 1`] = `
[
  {
    "fromClass": false,
    "methodName": "toUpperCase",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;

exports[`expression suggester should support safe navigation 1`] = `
[
  {
    "fromClass": false,
    "methodName": "quaxString",
    "refClazz": {
      "display": "String",
      "refClazzName": "java.lang.String",
    },
  },
]
`;
